src/components/Pagination/Pagination.tsx

```diff
@@ -1,42 +1,142 @@
+    selected: number;
 }
 
-export default function Pagination(
-    props: React.PropsWithChildren<PaginationProps>
-) {
-    const {
-        blockName,
-        children,
-        className,
-        modifiers,
-        nextPage,
-        previousPage,
-    } = props;
-
-    const pagination__base_class = "pagination";
-
-    return (
-        <nav
-            aria-label="Pagination"
-            className={bem(pagination__base_class, modifiers, blockName, [
-                className,
-            ])}
-        >
-            {previousPage}
-            {children}
-            {nextPage}
-        </nav>
-    );
+/** Pagination component that provides a list of page items */
+export default class Pagination extends React.Component<
+    PaginationProps,
+    PaginationState
+> {
+    constructor(props: PaginationProps) {
+        super(props);
+        this.state = {
+            selected: 0,
+        };
+    }
+
+    previousPage = () => {
+        const { selected } = this.state;
+        if (selected > 0) {
+            this.selectPage(selected - 1);
+        }
+    };
+
+    nextPage = () => {
+        const { selected } = this.state;
+        const { pageCount } = this.props;
+
+        if (selected < pageCount - 1) {
+            this.selectPage(selected + 1);
+        }
+    };
+
+    selectPage = (index) => {
+        if (this.state.selected === index) return;
+
+        this.setState({ selected: index });
+
+        // Call the callback with the new selected item:
+        this.runCallback(index);
+    };
+
+    runCallback = (selectedItem) => {
+        if (
+            typeof this.props.onPageChange !== "undefined" &&
+            typeof this.props.onPageChange === "function"
+        ) {
+            this.props.onPageChange({ selected: selectedItem });
+        }
+    };
+
+    getPageElement(index) {
+        const { selected } = this.state;
+
+        return (
+            <li key={index + 1}>
+                <Button
+                    onClick={this.selectPage.bind(null, index)}
+                    type="button"
+                    buttonType={ButtonTypes.Link}
+                    className={selected === index ? "selected" : ""}
+                >
+                    {index + 1}
+                </Button>
+            </li>
+        );
+    }
+
+    pagination = () => {
+        const items = [];
+
+        const { pageCount } = this.props;
+
+        for (let index = 0; index < pageCount; index++) {
+            items.push(this.getPageElement(index));
+        }
+
+        return items;
+    };
+
+    render() {
+        const pagination__base_class = "pagination";
+        return (
+            <nav
+                aria-label="Pagination"
+                className={bem(
+                    pagination__base_class,
+                    this.props.modifiers,
+                    this.props.blockName,
+                    [this.props.className]
+                )}
+            >
+                <ul>
```

Comment:
This is kind of a nit because I'm not sure it matters, ultimately, but `<ul>` doesn't seem semantically correct here.  You can't re-order the `<li>` items at will.  I'm seeing that Amazon is using `ul` anyway, while Google is using `table`, and both are weird to me.  I think I like 
https://react.semantic-ui.com/addons/pagination/
the best, where it's just a div with a bunch of links in it.  This is not tabular data, so table seems weird.  `ol` might be fine? Thoughts?